main.tf

```diff
@@ -68,13 +69,26 @@ resource "aws_transfer_server" "default" {
     }
   }
 
+  # this Terraform code block creates one or both of the on_upload and on_partial_upload blocks,
+  # depending on whether the corresponding input variables are defined and can be accessed.
   dynamic "workflow_details" {
-    for_each = local.on_upload
+    for_each = can(local.on_upload) || can(local.on_partial_upload) ? [1] : []
 
     content {
-      on_upload {
-        execution_role = var.on_upload.execution_role
-        workflow_id    = var.on_upload.workflow_id
+      dynamic "on_upload" {
+        for_each = can(local.on_upload) ? [1] : []
+        content {
+          execution_role = var.on_upload.execution_role
+          workflow_id    = var.on_upload.workflow_id
+        }
+      }
+
+      dynamic "on_partial_upload" {
+        for_each = can(local.on_partial_upload) ? [1] : []
+        content {
+          execution_role = var.on_partial_upload.execution_role
+          workflow_id    = var.on_partial_upload.workflow_id
+        }
       }
     }
   }
```

Comment:
Seeing as `workflow_details` is an optional attribute and it contains optional attributes, we don't need to create it as a dynamic block.

We also use the value of the local variable, `{ create = true }` or `{}`, as in the input for `for_each` so that we create either 1 or 0 instances of the dynamic block. Using `can()` always returns true in this instance (you can test this with `terraform console`):

```hcl
> can({})
true
```

With this in mind, the dynamic can be rewritten to the following:

```hcl
  workflow_details {
    dynamic "on_partial_upload" {
      for_each = local.on_partial_upload

      content {
        execution_role = var.on_partial_upload.execution_role
        workflow_id    = var.on_partial_upload.workflow_id
      }
    }

    dynamic "on_upload" {
      for_each = local.on_upload

      content {
        execution_role = var.on_upload.execution_role
        workflow_id    = var.on_upload.workflow_id
      }
    }
  }
```

Reply:
Thanks a lot for this feedback, I do learn a lot from it.